### DISCUSSION TITLE
Unable to boot Ubuntu 24.04

### DISCUSSION THREAD

**Discussion by vitali-fridman** [Sithuk/ubuntu-server-zfsbootmenu#48] [General]
I used this scrips many times with Ubuntu 22.04, 23.03 and 23.10. I'm now trying it with 24.04 and there is a problem. Script runs without error, but system does not boot. The boot starts (it outputs first few traces) and the host immediately reboots and it's a loop. I can import the rpool into live environment without issues. Also installing from the desktop installer with built-in ZFS option works, but this obviously not what I want as it's only for single disk. How can I troubleshoot this more?

**Comment by Sithuk** [comment]
There have been a few reported issues with booting. Check the other recent discussion threads, but one user noted that their issue was within the BIOS/UEFI and the resolution involved loading factory default secure boot keys.

**Comment by vitali-fridman** [comment]
I have default secure boot keys. I resolved the issue by installing 22.04 and upgrading it to 24.04. That worked.

**Comment by almightiest** [comment]
In my 24.04 initial install post-reboot, I had to go into zfsbootmenu and import the new rpool manually.  I also had to export my other pools as they were imported into zfsbootmenu instead (i had backups of other os's with rpools and i think it got confused).  It worked from there on out.

**Reply by vitali-fridman** [reply]
Unfortunately I'm not super familiar with zbm workings. Do you mean that you import correct rpool before the actual boot sequence starts? Can you elaborate how to do that? I imagine you need some sort of command line or maybe in some other way?

**Reply by almightiest** [reply]
Basically I had to drop into the ZBM command line, and run a few commands (i think these are right):

`zpool list` to see what's imported
`zpool export {anything in that list that's not the new rpool}`
`zpool import` to find a list of pools available to import
`zpool import -d /dev/disk/by-id {correct rpool}`

**Comment by vitali-fridman** [comment]
Is dropping to ZBM command shell done with [Control+R]?

**Comment by vitali-fridman** [comment]
That solution did not work for me. However I saw an error message when I dropped to ZBM command shell. Correct zpool was imported by the error message says that linux kernel was not found.

I realized that this is probably bogus message as I've seen it in many case when the boot proceeds correctly.

**Comment by vitali-fridman** [comment]
I still don't understand what could be wrong. By now I installed 24.04 on a dozen machines with variety of hardware configurations and I can't boot direct install of 24.04, needed to do an upgrade from 22.04 on all of them.

**Comment by Sithuk** [comment]
Are you able to post the installation log? Remember to remove any sensitive details such as your password.

**Comment by vitali-fridman** [comment]
[ubuntu_setup_zfs_root.log](https://github.com/user-attachments/files/16147801/ubuntu_setup_zfs_root.log)

Log attached, it's from the latest version of the script. When rebooting after running "initial" target: zfsbootmenu works, boot starts, couple of normal first traces appear on a console (something like "Tearing down USB hub") and immediately system goes into reset/reboot.

**Reply by Sithuk** [reply]
I haven't spotted anything obviously incorrect in your log. I see you are using an nvme drive. I haven't tested with one of those. Could there be a driver/feature that isn't being installed for an nvme drive?

**Reply by vitali-fridman** [reply]
I'm not an expert but my guess nvme driver should be included in standard Ubuntu kernel build. I will try to install on external USB SSD and see if that makes a difference.

**Reply by vitali-fridman** [reply]
I install to external USB SSD and behavior is exactly the same. I also did not make any changes to original script except user name/password, host name and pool name and locale/time zone, increased timeout to 30 sec and disable encryption. My guess that boot fails trying to initialize the hardware. But for me it happen on machines with very different hardware - CPU, MB, RAM all different. The only common thing is that CPUs are all from AMD: 5950X, 3990X, 3995WX. 3970X. And motherboards are all from ASUS, but they are of three different types. Maybe something about combination of ASUS motherboards/UEFI and AMD CPUs make it happens? Does not make much sense. 22.04 installs fine with the same script and Ubuntu 24.04 single disk desktop install (so called experimental feature in standard installer) is fine.

**Reply by vitali-fridman** [reply]
I wander if standard 24.04 installer from Canonical has different version of kernel that the script pulls from the apt repositories. Right now the script installs 6.8.0-36 and I suspect that standard installer image has version 6.8.0-35 or even prior. Just a guess.

**Reply by Sithuk** [reply]
What does the system journal report following the failed boot? You can enter a chroot environment from zfsbootmenu. Run journalctl from the chroot environment.

**Reply by vitali-fridman** [reply]
journalctl returns: "No journal files were found. No entries"

**Reply by Sithuk** [reply]
Anything in syslog in /var/log or other log files in there?

**Reply by vitali-fridman** [reply]
There are no files in /var/log.

**Reply by vitali-fridman** [reply]
I noticed that during successful boot the traces that appear in that spot are something like "Loaded microcode...".  These traces do not appear during failed boot. I believe this is loading to CPU and that what triggers system reset.

**Reply by vitali-fridman** [reply]
I made some progress. The machine I've been doing this last couple of days had somewhat old BIOS. I updated it to the latest and the problem looks to be resolved. My guess is that the latest greatest Ubuntu build has AMD microcode that is incompatible with some older BIOSes. But the build which is in live environment ISO is compatible. However it also had this problem on some machines with another type of CPU/MB that I believe have the latest BIOS, but I need to double check on that.

**Reply by vitali-fridman** [reply]
I can confirm that this problem is resolved by updating BIOS to the latest. This seems to only affect AMD based systems but it seems to affect all kinds of AMD CPUs. Also zfsbootmenu that is build while installing 24.04 can't mount pools created in 22.04 installation that has not been upgraded  via do-release-upgrade. Also pre-24.04 versions of zfsbootmenu can't mount 24.04 created pools, as they have some new features. I speculate that upgrading pools in upgraded 24.04 system without upgrading zfsbootmenu first will result of unbutable system.

**Reply by Sithuk** [reply]
"pre-24.04 versions of zfsbootmenu can't mount 24.04 created pools, as they have some new features"
The zfs version and kernel module in the zfsbootmenu image files will determine zfs pool support. The zfs version and kernel module in a user compiled zfsbootmenu version is determined by the zfs version and kernel module on the system that zfsbootmenu was compiled on. I haven't tested it, but you can try updating zfs and the kernel on a 22.04 system and then run "sudo generate-zbm" to update the zfsbootmenu image files. The zfs and kernel module in Zfsbootmenu should then be updated. You can check what zfs version is installed in zfsbootmenu by entering the zfsbootmenu recovery console and entering "zfs --version".

"zfsbootmenu that is build while installing 24.04 can't mount pools created in 22.04 installation"
That's odd. 24.04 currently has zfs version 2.2.2 which is more recent than the zfs version 2.1.5 currently in 22.04. If zfsbootmenu was compiled in 24.04, then it should be able to load a pool created in 22.04.

**Reply by vitali-fridman** [reply]
Sorry, my mistake. Yes, you can import pools created in 22.04 from 24.04-build zfsbootmenu. However it did not import these pools by default. I had to go into recover console and import them with "-f" force flag.

**Reply by almightiest** [reply]
> Sorry, my mistake. Yes, you can import pools created in 22.04 from 24.04-build zfsbootmenu. However it did not import these pools by default. I had to go into recover console and import them with "-f" force flag.

That's the same thing I had to do after first reboot as well. For what it's worth, I had other zfs data pools and drives that may have confused the installer.
